module/os/linux/zfs/abd_os.c

```diff
@@ -1014,6 +1100,115 @@ abd_cache_reap_now(void)
 {
 }
 
+/*
+ * Borrow a raw buffer from an ABD without copying the contents of the ABD
+ * into the buffer. If the ABD is scattered, this will allocate a raw buffer
+ * whose contents are undefined. To copy over the existing data in the ABD, use
+ * abd_borrow_buf_copy() instead.
+ */
+void *
+abd_borrow_buf(abd_t *abd, size_t n)
+{
+	void *buf;
+	abd_verify(abd);
+
+	/*
+	 * In the event the ABD is composed of a single user page from Direct
+	 * I/O we can not direclty return the raw buffer. This is a consequence
+	 * of not being able to write protect the page and the contents of the
+	 * page can be changed at any time by the user.
+	 */
+	if (abd_is_from_pages(abd)) {
+		buf = zio_buf_alloc(n);
```

Comment:
This workaround looks weird to me.  As I understand it, you are protecting code paths that are unable to use scatter ABDs directly.  So good algorithms are allowed to suffer, while bad ones are protected.  Am I missing something?

Reply:
This came out of the same OpenZFS Leadership Meeting as the percentage check module parameter. So the issue here is that with Linux anonymous pages cannot be put under write protection. So if compression is turned on and there is a linear page then the user can mess with that all day and cause data corruptions or an explosion when doing the compression algorithm.

Reply:
Then it is not a direct I/O.

Reply:
Not Direct I/O in the traditional sense. It is important to note that even without the linear page any abd that is `ABD_FLAG_FROM_PAGES` is going to be copied by `and_borrow_buf_copy()` as they are normally scatter ABD's. This is a consequence of the compression needing a linear buffer. So, a copy will be made with a Direct I/O buffer if compression is enabled. This is also why LANL was working on improving this with the Z.I.A. PR (#13628). While I was working on Direct I/O, @calccrypto was working on the other part of the equation to improving ZFS performance with NVMe devices.

I am not aware of a way to avoid doing the copy here. I am sure at some point a patch could maybe be created to handle compression with scatter ABD's. However, that should probably be done in a different PR. There still would be the problem in the meantime of Linux now allowing us to write protect anonymous user pages. So the only safe way to do compression in that case is through a copy. I have not looked recently at the latest kernel code, so it is maybe the case they now allow anonymous pages to be placed under write protection. I doubt that though. Those are user pages and the kernel in general does no care what a user does with those pages.